Support repeats in Google Sheets

grzesiek2010 opened the pull request:
Closes #1385 

#### What has been done to verify that this works as intended?
I tested many different forms:
A form with one repeat,
A form with multiple repeats,
Media files in repeats,
Empty answers etc.

#### Why is this the best possible solution? Were any other approaches considered?
It works in a similar way like Briefcase and Aggregate - repeats are stored in a separate sheet.

#### Are there any risks to merging this code? If so, what are they?
Of course, uploading to Google sheets should be tested carefully.

#### Do we need any specific form for testing your changes? If so, please attach one.
Any form with a repeat section.

grzesiek2010 committed:
Extracted areHeadersEmpty() method

grzesiek2010 committed:
Extracted areColumnNamesLegal() method

grzesiek2010 committed:
Extracted areWritePermissionsGranted() method

grzesiek2010 committed:
Extracted method uploadMedia()

grzesiek2010 committed:
Extracted resizeSpreadSheet() method

grzesiek2010 committed:
Extracted addHeaders() method

grzesiek2010 committed:
Extracted isColumnLengthValid() method

grzesiek2010 committed:
Extracted readColumnNames method()

grzesiek2010 committed:
Aded instanceFile as a method parameter

grzesiek2010 committed:
Extracted readAnswers() method

grzesiek2010 committed:
Extracted areSubmissionColumnNamesLegal() method

grzesiek2010 committed:
Extracted insertRow method()

grzesiek2010 committed:
Extracted areEmptyColumns() method

grzesiek2010 committed:
Extracted handleBlankColumnNames() method

grzesiek2010 committed:
Extracted addMissingColumns() method

grzesiek2010 committed:
Extracted checkForMissingColumns() method

grzesiek2010 committed:
Extracted addPhotos() method

grzesiek2010 committed:
Extracted updateValues() method

grzesiek2010 committed:
Extracted getSheetCols() method

grzesiek2010 committed:
Extracted prepareListOfValues() method

grzesiek2010 committed:
Extracted getRowFromList() method

grzesiek2010 committed:
Use insertRow() method in other places too

grzesiek2010 committed:
Extracted sleepThread() method

grzesiek2010 committed:
Removed unnecesarry comments

grzesiek2010 committed:
Reduced the code using exceptions

grzesiek2010 committed:
Fixed bug in updateValues() method

grzesiek2010 committed:
Removed unnecessary areHeadersEmpty() method

grzesiek2010 committed:
Removed unnecessary code

grzesiek2010 committed:
Improved naming

grzesiek2010 committed:
Moved token check up

grzesiek2010 committed:
Improved comments

grzesiek2010 committed:
Implemented addSheet() method

grzesiek2010 committed:
Support repeat groups

grzesiek2010 committed:
Code improvements

grzesiek2010 committed:
Fixed token check

grzesiek2010 committed:
Removed unnecessary validation

grzesiek2010 committed:
Fixed problem with empty groups

grzesiek2010 committed:
Fixed problem with doubled submissions

grzesiek2010 committed:
Fixed problem with empty answers

grzesiek2010 committed:
Fixed naming

grzesiek2010 committed:
Fixed lints

shobhitagarwal1612 commented:
Form used for testing : `Forest Plot Survey`

-  4 form instances were uploaded with repeats.
     - 1st form  - 3 repeats
     -  2nd form - 3 repeats (same as 1st)
     -  3rd form - 2 repeats (with blank data)
     -  4th form - 1 repeat


Observations : 
-  For each row in the non-repeat group, the column with the start of repeat is appearing twice 
    (1st image) 
-  As you can see in the image attached below, for each submission the repeats are uploaded twice. The number of rows with data in the second image should only be 9 but they are actually 18. 
    (2nd image)
-  How will one be able to recognize the repeats corresponding to a particular row after it is uploaded? For example, if I had not mentioned above the number of repeats for each submission then there was no way to find that out by looking at the sheet.

![image](https://user-images.githubusercontent.com/8918023/36169473-51985bca-1122-11e8-9b72-86a8b3574a8b.png)

![image](https://user-images.githubusercontent.com/8918023/36169512-6a044e6c-1122-11e8-8d4a-3d68ceea707e.png)

grzesiek2010 committed:
Fixed bug with duplicated repeat groups

grzesiek2010 commented:
@shobhitagarwal1612 I fixed the issue.

shobhitagarwal1612 commented:
> For each row in the non-repeat group, the column with the start of repeat is appearing twice
(1st image)
As you can see in the image attached below, for each submission the repeats are uploaded twice. The number of rows with data in the second image should only be 9 but they are actually 18.
(2nd image)

I can verify that both of the above issue are fixed.

> How will one be able to recognize the repeats corresponding to a particular row after it is uploaded? For example, if I had not mentioned above the number of repeats for each submission then there was no way to find that out by looking at the sheet.

Can we link the cell from the first sheet to the second sheet's cell? In that way we will know that which repeat belongs to which row in the first sheet. 
Something like `SheetName!CellAddress` (eg `Sheet1!B5`)

lognaturel commented:
> Can we link the cell from the first sheet to the second sheet's cell?

This is really important. I think we should match what Briefcase does -- include the `instanceId` in the parent sheet and a corresponding `parentId` in the repeat sheets. I believe this always works with Briefcase because if a form doesn't include an `instanceId`, Aggregate generates one. In this case, perhaps a form with repeats and no `instanceId` should be rejected.

I believe what is going on now is that the first question is being used as a link, is that right, @grzesiek2010? That will work in most cases if `instanceId` is part of the form design but will fail in cases like the forest plot form where no instanceId is specified.

grzesiek2010 commented:
> I believe what is going on now is that the first question is being used as a link, is that right

yes and I agree that it's important but `Forest Plot Survey` is a specific form since it doesn't contain `instanceID`. I noticed that creating forms using `xls` or [ODK Build](https://build.opendatakit.org/) an `instanceID` is always there.

> Something like SheetName!CellAddress (eg Sheet1!B5)

I think it might be tricky because someone might delete something manually and then such a link might be wrong and another thing is that Aggregate and Briefcase use instanceId for that so I think my current solution is ok. Maybe we should just check if an `instanceId` exists and don't send such a form if not.

shobhitagarwal1612 commented:
> I think it might be tricky because someone might delete something manually and then such a link might be wrong and another thing is that Aggregate and Briefcase use instanceId for that so I think my current solution is ok. Maybe we should just check if an instanceId exists and don't send such a form if not.

Agreed 👍

grzesiek2010 commented:
@lognaturel so what should I do? Leve it as it is or add this?:

> Maybe we should just check if an instanceId exists and don't send such a form if not.

lognaturel commented:
> I noticed that creating forms using xls or ODK Build an instanceID is always there.

Correct. XLSForm puts it at the end of the form and in general, there is no guarantee about the location of `instanceID` in the form. You should explicitly use the `instanceID` question as the link no matter where in the form it's located. Forms without an `instanceID` should not be allowed for sheets.

I have tried two forms that fail to upload the main form body. Some of the repeats are uploaded but not all. [groups-repeats.xml.txt](https://github.com/opendatakit/collect/files/1734045/groups-repeats.xml.txt) fails with "No data found" or "Sorry, no form was uploaded" -- I wasn't able to quickly tell what explains why it's not a consistent error. All student names repeats were uploaded but not all student grades. The other form which failed I unfortunately can't share but the failure mode seems to be the same. I can see that there's an `ArrayOutOfBounds` exception thrown at `answers.put(columnTitles.get(j), Iterables.get(answersToUpload.get(columnTitles.get(j)), j == 0 ? 0 : i))` but I haven't yet had a chance to look beyond that. The stacktrace gets swallowed by the `Exception` catch-all which is a problem. In general, exceptions should be handled as locally as they possibly can. The new structure that uses `Exception`s for control flow violates that.

At a high-level, I recommend spending some time looking for opportunities to make the code more readable and maintainable. That line above is a great example where it's unclear what is going on. A combination of additional comments and breaking 7b17a46 with commit messages that explain some of the decisions would also help a lot. For example, is using a guava `Multimap` really the best option or could you make the code easier to follow by keeping your own lists of maps? It should be immediately clear from looking at the code that the first question (soon to be changed to `instanceID`) is used to link the main form to repeats. It's not necessarily clear what `elements.add(0, mainLevelColumnElements.get(0));` is for. Similarly, `fixBlankColumnNames` really should have a comment linking to the issue that it's addressing since now that it's moved it'll be harder to get to the commit that added it.

There are a few methods that take in blank collections and then populate them that should instead return (https://books.google.com/books?id=_i6bDeoCQzsC&pg=PA45&lpg=PA45&dq=Output+arguments+should+be+avoided&source=bl&ots=ep3SFo9d46&sig=GkBX61M8oLwgh6wrxx0BBtgKqU0&hl=en&sa=X&ved=0ahUKEwjz2JLo3q3ZAhUOSK0KHZ_TAkQ4ChDoAQgoMAE#v=onepage&q=Output%20arguments%20should%20be%20avoided&f=false). A clear example is `uploadMedia` -- it should not pass have a parameter for `uploadedMedia` but should build the collection and return it. There are very short methods like `addMediaFiles` that make things harder to follow and could simply stay inline. There are some others that are a little less clear what to do with because currently they populate multiple collections. You can see if you come up with clearer ways of writing them and I will take another look as well.

Breaking `uploadOneSubmission` down into methods has brought up the challenge of how to propagate errors. I see you first tried to use boolean returns and then ended up using generic `Exception` throwing. I think this is a structure that should generally be avoided -- http://wiki.c2.com/?DontUseExceptionsForFlowControl. It also causes problems when there are unexpected exceptions like the `ArrayOutOfBounds` exception I described above.

grzesiek2010 committed:
Check for instanceID

grzesiek2010 committed:
Support nested repeat groups

grzesiek2010 committed:
Removed usning Exceptions for flow control

grzesiek2010 committed:
Code improvements

grzesiek2010 committed:
Removed unnecessary check - Google Sheets allow adding other chars too

grzesiek2010 committed:
Code improvements

grzesiek2010 committed:
Improved previous solution - read answers from treeElements insead of parsing submission file

grzesiek2010 committed:
Ignore empty rows

grzesiek2010 committed:
Code improvements

grzesiek2010 committed:
Fixed max columns limit - it's 256 not 255

grzesiek2010 committed:
Fixing blank headers didn't work

grzesiek2010 committed:
Ignore template elements

grzesiek2010 commented:
@dcbriccetti could you review the code?

I changed almost everything in comparison with the old version of `InstanceGoogleSheetsUploader` class.
I started with code refactoring since the method responsible for uploading submissions had 400 lines.
Then I implemented changes for uploading repeat groups.
After that, I noticed that the way we use for reading column names and answers (parsing XML file) was really difficult to maintain and read (especially with changes for repeats), so I refactored that functionality.

In the end, I investigated everything again and I noticed that some functions don't work and some are not needed:

- I removed checking if uploaded strings are valid https://github.com/opendatakit/collect/blob/master/collect_app/src/main/java/org/odk/collect/android/tasks/InstanceGoogleSheetsUploader.java#L99 since I tested not allowed chars and we can use it. Probably it was not allowed in old Google Sheets API.

- I fixed inserting column headers again when any blank header is detected https://github.com/opendatakit/collect/blob/master/collect_app/src/main/java/org/odk/collect/android/tasks/InstanceGoogleSheetsUploader.java#L466 (it didn't work)

- I also removed [Thread.sleep(GOOGLE_SLEEP_TIME)](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/java/org/odk/collect/android/tasks/InstanceGoogleSheetsUploader.java#L262) since I uploaded lots of different forms without this line and I didn't notice we need that, probably it was needed in old Google Sheet API after my big changes I don't even know where it could be added.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/tasks/InstanceGoogleSheetsUploader.java:
```diff
@@ -556,242 +519,123 @@ private boolean uploadOneSubmission(String id, String instanceFilePath, String j
     }
 
-    public void setAuthFailed() {
+    public void setAuthFailedForFalse() {
```
`setAuthFailedToFalse`

dcbriccetti replied:
Or `unsetAuthFailed` or `resetAuthFailed` or `setAuthFailed(boolean authFailed)`

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/tasks/InstanceGoogleSheetsUploader.java:
```diff
@@ -199,12 +199,19 @@ private boolean uploadOneSubmission(String id, File instanceFile, String jrFormI
             FormDef formDefFromXml = XFormUtils.getFormFromInputStream(new FileInputStream(new File(formFilePath)));
 
             List<TreeElement> mainLevelColumnElements = getColumnElements(formDefFromXml.getMainInstance().getRoot());
+            TreeElement instanceIDColumn = getInstanceIDColumn(mainLevelColumnElements);
```
Can you get this from the database to avoid doing a full pass through the form? It's a small optimization but could make a difference when a form has <strike>thousands of questions</strike>255 questions.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/tasks/InstanceGoogleSheetsUploader.java:
```diff
@@ -113,23 +112,20 @@ public static boolean isLocationValid(String answer) {
     }
 
-    private String getGoogleSheetsUrl(Cursor cursor) {
+    private String getSpreadsheetUrl(Cursor cursor) {
```
`getGoogleSeetsUrl` seems more accurate!

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/tasks/InstanceGoogleSheetsUploader.java:
```diff
@@ -152,399 +186,328 @@ private void uploadInstances(String selection, String[] selectionArgs, String to
-                    if (!uploadOneSubmission(id, instance, jrformid, token, formFilePath, urlString)) {
-                        cv.put(InstanceColumns.STATUS,
-                                InstanceProviderAPI.STATUS_SUBMISSION_FAILED);
+                    if (token == null) {
```
Seems `token` doesn't change in the method so this check should come before any other work is done.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/tasks/InstanceGoogleSheetsUploader.java:
```diff
@@ -152,399 +186,328 @@ private void uploadInstances(String selection, String[] selectionArgs, String to
+            TreeElement instanceIDElement = getInstanceIDElement(getChildElements(instanceElement));
+            if (hasRepeatableGroups(instanceElement)) {
+                if (instanceIDElement == null) {
+                    outcome.results.put(id, "This form contains repeatable group so it should contain an instanceID!");
```
This will show up in the UI so it should be translated. More concise version: "Forms with repeats must define an instanceID."

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/tasks/InstanceGoogleSheetsUploader.java:
```diff
@@ -152,399 +186,328 @@ private void uploadInstances(String selection, String[] selectionArgs, String to
-            Timber.e(e2, "Exception thrown while getting columns from form file");
-            outcome.results.put(id, e2.getMessage());
+            insertRows(instanceElement, instanceIDElement, instanceFile, mainSheetTitle);
+        } catch (Exception e) {
```
This is still using exceptions for flow control.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/tasks/InstanceGoogleSheetsUploader.java:
```diff
@@ -152,399 +186,328 @@ private void uploadInstances(String selection, String[] selectionArgs, String to
-        if (columnNames.size() > 255) {
-            outcome.results.put(id, Collect.getInstance().getString(R.string.sheets_max_columns,
-                    String.valueOf(columnNames.size())));
+    private boolean insertRow(TreeElement element, TreeElement instanceIDElement, File instanceFile, String sheetTitle) throws Exception {
```
Either the return value should be used or this should be void.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/tasks/InstanceGoogleSheetsUploader.java:
```diff
@@ -556,242 +519,123 @@ private boolean uploadOneSubmission(String id, String instanceFilePath, String j
             Timber.e(e);
             outcome.results.put(id, e.getMessage());
-            return false;
+            throw e;
```
Exception for control flow. This is a small enough method that I think it would be more clear to include it inline.

I know this is existing code but `getHeaderFeed` doesn't seem like the right name for that method, right? Seems **it** should be `getSheetCells`.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/tasks/InstanceGoogleSheetsUploader.java:
```diff
@@ -152,399 +186,328 @@ private void uploadInstances(String selection, String[] selectionArgs, String to
-            if (!sheetCols.contains(col)) {
-                missingColumns.add(col);
-            }
+    private String getElementTitle(AbstractTreeElement element) {
```
This should get a comment describing that this builds a column name by joining all of the containing group names using "-" as a separator.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/tasks/InstanceGoogleSheetsUploader.java:
```diff
@@ -556,242 +519,123 @@ private boolean uploadOneSubmission(String id, String instanceFilePath, String j
-                }
+    private boolean isNumberOfColumnsValid(int numberOfColumns) {
+        if (numberOfColumns == 0) {
+            outcome.results.put(id, "No columns found in the form to upload");
```
Should be translated.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/tasks/InstanceGoogleSheetsUploader.java:
```diff
@@ -152,399 +186,328 @@ private void uploadInstances(String selection, String[] selectionArgs, String to
+    private List<Object> getColumnTitles(TreeElement element, TreeElement instanceIDElement) {
+        List<Object> columnTitles = new ArrayList<>();
+        if (element.isRepeatable()) {
+            columnTitles.add(0, getElementTitle(instanceIDElement));
```
Again, using the `INSTANCE_ID` constant would be clearer.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/tasks/InstanceGoogleSheetsUploader.java:
```diff
@@ -152,399 +186,328 @@ private void uploadInstances(String selection, String[] selectionArgs, String to
+        HashMap<String, String> answers = getAnswers(getChildElements(element), instanceFile);
+        if (shouldRowBeInserted(answers)) {
+            if (element.isRepeatable()) {
+                answers.put(getElementTitle(instanceIDElement), instanceIDElement.getValue().getDisplayText());
```
`getElementTitle(instanceIDElement)` will always be `FormController.INSTANCE_ID`, right? That would be clearer.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/tasks/InstanceGoogleSheetsUploader.java:
```diff
@@ -152,399 +186,328 @@ private void uploadInstances(String selection, String[] selectionArgs, String to
-            ValueRange row = new ValueRange();
-            row.setValues(content);
+    private List<Object> getColumnTitles(TreeElement element, TreeElement instanceIDElement) {
+        List<Object> columnTitles = new ArrayList<>();
```
Why isn't this a `List` of `String`s?

grzesiek2010 replied:
Because a list of objects is used in the API to insert rows so I use the same here to avoid converting.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/tasks/InstanceGoogleSheetsUploader.java:
```diff
@@ -71,78 +80,103 @@
-    private String googleSheetsUrl = "";
+
+    private String mainSheetTitle;
+    private String googleSheetsUrl;
```
You should always be able to get this from a `Spreadsheet` object.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/tasks/InstanceGoogleSheetsUploader.java:
```diff
@@ -71,78 +80,103 @@
     private boolean authFailed;
-    private String googleSheetsUrl = "";
+
+    private String mainSheetTitle;
```
You should always be able to get this from a `Spreadsheet` object. Currently it looks like it's only used in `insertRows` so it can be computed locally.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/tasks/InstanceGoogleSheetsUploader.java:
```diff
@@ -71,78 +80,103 @@
+
+    private String mainSheetTitle;
+    private String googleSheetsUrl;
     private String spreadsheetId;
```
You should always be able to get this from a `Spreadsheet` object.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/tasks/InstanceGoogleSheetsUploader.java:
```diff
@@ -152,399 +186,328 @@ private void uploadInstances(String selection, String[] selectionArgs, String to
+        return true;
+    }
 
+    private void insertRow(List<Object> rowElements, String sheetName) throws IOException {
```
I don't think this method is helpful. If a method name is overloaded, it should be to have different variants with different parameter lists. Here, the two `insertRow` methods are not related in this way. Making the helper call and catching the exception in the primary `insertRow` would be clearer.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/tasks/InstanceGoogleSheetsUploader.java:
```diff
@@ -152,399 +186,328 @@ private void uploadInstances(String selection, String[] selectionArgs, String to
+            } else {
+                String answer = element.getValue() != null ? element.getValue().getDisplayText() : "";
+                if (new File(instanceFile.getParentFile() + "/" + answer).isFile()) {
+                    answers.put(elementTitle, uploadMediaFile(instanceFile, answer));
```
It's easy to miss that the attachment is actually uploaded here. Clearer:
```
String mediaHyperlink = uploadMediaFile(instanceFile, answer);
answers.put(elementTitle, mediaHyperlink);
```

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/tasks/InstanceGoogleSheetsUploader.java:
```diff
@@ -71,78 +80,103 @@
+    private String mainSheetTitle;
+    private String googleSheetsUrl;
     private String spreadsheetId;
+    private String id;
```
`id` is only used for the outcome text. Could `jrFormId` be used in the outcome instead so that this field isn't needed?

dcbriccetti replied:
I’ve eliminated it in my proposed changes.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/tasks/InstanceGoogleSheetsUploader.java:
```diff
@@ -71,78 +80,103 @@
     private Outcome outcome;
-    private boolean hasWritePermissionToSheet = false;
+
+    private boolean hasWritePermissionToSheet;
```
Is this field actually useful? It's only ever set to true so I think that means even if you're uploading a bunch of different forms to different sheets, as soon as you hit the first one you have write access to the field is true. I think it can be removed.

lognaturel reviewed (changes requested):
There are a lot of really good improvements here. I've provided some more suggestions in line. In particular, it would be great to reduce the amount of state kept to keep things easier to reason about.

There are still a number of methods that throw exceptions and a `catch (Exception e)` catch-all in `uploadOneInstance`. @dcbriccetti do you have any best practices to share in this case? The scenario is that a task has multiple subtasks and the failure of any one of them should result in the failure of the overall task. Uploading one instance entails making sure the spreadsheet is set up to receive that instance, gathering all the relevant answers, writing those answers to the spreadsheet, etc. A lot of the calls involved can throw exceptions. Remove the `throws Exception` in `insertRow` to see that basically every single line can throw an exception. We don't particularly care about those exceptions and just want to stop (though in an ideal world we would clean up the partial submission).

The original solution included giant methods with lots of try-catches.

dcbriccetti commented:
>@dcbriccetti could you review the code?

Sorry, I missed this way back when. One comment right away is that many of the commits could be combined, and that would help reviewers.

dcbriccetti commented:
Here are the changes I recommend, to simplify the error handling:

https://github.com/dcbriccetti/collect/commit/fd36388ef0ffcdb91c093ff1003b413da4d1fd5e

grzesiek2010 committed:
Code improvements

grzesiek2010 commented:
@lognaturel @dcbriccetti I implemented all improvements.

lognaturel commented:
Thanks for getting those changes in so quickly, @grzesiek2010! @dcbriccetti to summarize, you're not against using exceptions for control flow in a situation like this but introducing a custom exception means that at least we can be sure that it's an "expected" exception. Did I interpret your suggestions right?

I think this is now clear enough that we can move to QA. @dcbriccetti?

dcbriccetti commented:
Sure, we can move to QA.

If this were in Scala (or possibly with more modern Java features available), I would probably not recommend the exceptions. (Scala has `Either` which allows a function to return either a normal value or an object that contains details about the error.) 

What I suggested creates a single exception 
`/** An exception that results in the cancellation of an instance upload, and the presentation of an error to the user */`
that encapsulates either a message string or an exception (cause).
It keeps the IOException and other specific exceptions down in the lower levels.

Here’s my commit comment, since it didn’t survive:

Improve InstanceGoogleSheetsUploader error handling

Rename Outcome.results to messagesByInstanceId
Eliminate fields `outcome` and `id`
Have lower level methods not modify `outcome`, but rather return errors via a new exception, `UploadException`
Remove some duplicate code
In-line `resizeSheet`
Rename and change `isAnyMissingColumn` and `isNumberOfColumnsValid` to throw UploadException

dcbriccetti commented on collect_app/src/main/java/org/odk/collect/android/tasks/InstanceGoogleSheetsUploader.java:
```diff
@@ -68,483 +78,418 @@
-            if (!sheetCols.contains(col)) {
-                missingColumns.add(col);
-            }
+    // This  method builds a column name by joining all of the containing group names using "-" as a separator
```
We might as well enjoy the benefits of having this be a doc comment. Pop-up help in the IDE, for instance.
`/** This method builds a column name by joining all of the containing group names using "-" as a separator */
(Please notice the removed extra space.)

dcbriccetti commented on collect_app/src/main/java/org/odk/collect/android/tasks/InstanceGoogleSheetsUploader.java:
```diff
@@ -556,242 +501,104 @@ private boolean uploadOneSubmission(String id, String instanceFilePath, String j
 
-                // check if root folder exists, if not then create one
-                driveHelper.getIDOfFolderWithName(GOOGLE_DRIVE_ROOT_FOLDER, null, true);
+    private String getGoogleSeetsUrl(Cursor cursor) {
```
`Sheets`, not `Seets`

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/activities/InstanceUploaderActivity.java:
```diff
@@ -167,7 +167,7 @@ protected void onDestroy() {
     @Override
     public void uploadingComplete(HashMap<String, String> result) {
-        Timber.i("uploadingComplete: Processing results (%d) from upload of %d instances!",
+        Timber.i("uploadingComplete: Processing messagesByInstanceId (%d) from upload of %d instances!",
```
This looks like a search and replace error.

grzesiek2010 committed:
Code improvements

grzesiek2010 commented:
@mmarciniak90 I've added some final changes it's not important and shouldn't change anything but please update your branch for testing just in case.

mmarciniak90 commented:
**Regression has been found!**

I checked this case on v1.13.2. Problem was not visible.

**Steps to reproduce:**
1. Start form with video - [one video.xml.txt](https://github.com/opendatakit/collect/files/1819734/one.video.xml.txt)
2. Add video
3. Send form to Google Sheets (I added a link to sheets to Fallback submission URL)

**Current behavior:**
* Error appears 
<img height="40%" width="40%" src="https://user-images.githubusercontent.com/3907554/37529092-208313c6-292e-11e8-846a-eca9e0d872b9.png">
but the file exists in directory
<img height="40%" width="40%" src="https://user-images.githubusercontent.com/3907554/37529116-327c6460-292e-11e8-85f5-c2f6d8f2987e.png">

grzesiek2010 committed:
Fixed problem with sending media files

grzesiek2010 commented:
@mmarciniak90 I fixed the problem and you can continue testing once you update your branch.

dcbriccetti commented on collect_app/src/main/java/org/odk/collect/android/tasks/InstanceGoogleSheetsUploader.java:
```diff
@@ -274,20 +273,11 @@ private boolean shouldRowBeInserted(HashMap<String, String> answers) {
-                        selectionArgs, null);
-        if (cursor != null && cursor.getCount() != 1) {
-            cursor.close();
+        if (!new File(filePath).exists()) {
```
To help reviewers, would you please explain this change and how it fixes the problem? Thanks.

mmarciniak90 commented:
**Tested with success!**

Verified on Android: 4.1, 4.2, 5.1, 6.0, 7.0, 8.0

Verified cases:
* few looped repeat groups
* one repeat group
* form with photos, vidoes and audio files in repeat group
* empty answer
* empty repeat group
* form with images in multi select widget 
* regression on aggregate
* removing groups
* Google Drive permissions:
<img height="40%" width="40%" src="https://user-images.githubusercontent.com/3907554/37650364-a16c76d4-2c2c-11e8-9426-46e836d0cd31.png">

@opendatakit-bot unlabel "needs testing"
@opendatakit-bot label "behavior verified"

lognaturel commented:
👏 👏 👏
